packages/vite/src/client/error.ts

```diff
@@ -0,0 +1,266 @@
+import type { ErrorPayload } from 'types/hmrPayload'
+
+export const transformError = (error: any): Error => {
+  if (error instanceof Error) {
+    return error
+  }
+
+  const e = Error(error ?? 'unknown')
+  e.stack = `a non-error was thrown please check your browser's devtools for more information`
+  return e
+}
+
+const RE_CHROME_STACKTRACE =
+  /^ {4}at (?:(.+?)\s+)?\(?(.+?)(?::(\d+))?(?::(\d+))?\)?$/
+const RE_FIREFOX_STACKTRACE =
+  /^(?:(?:(^|.+?)@))\(?(.+?)(?::(\d+))?(?::(\d+))?\)?$/
+
+export interface StackLineInfo {
+  input: string
+  varName: string
+  url: string
+  line: number
+  column: number
+  vendor: 'chrome' | 'firefox'
+}
+
+export type StackLineResult = Partial<StackLineInfo> & { input: string }
+
+export const isStackLineInfo = (res: StackLineResult): res is StackLineInfo => {
+  return res.url !== undefined
+}
+
+export const getStackLineInformation = (line: string): StackLineResult => {
+  let match = RE_CHROME_STACKTRACE.exec(line)
+  if (match) {
+    let [input, varName, body, line, column] = match
+    // strip eval
+    if (body.indexOf('eval at') !== -1) {
+      body = body.replace(/(eval at.*\()|(\),.*)/g, '')
+      const lcMatch = /(.*):(\d+)?:(\d+)?$/.exec(body)
+      if (lcMatch) {
+        body = lcMatch[1]
+        line = lcMatch[2]
+        column = lcMatch[3]
+      } else {
+        // can't extract column
+        return {
+          input
+        }
+      }
+    }
+
+    return {
+      input,
+      varName,
+      url: body,
+      line: Number(line),
+      column: Number(column),
+      vendor: 'chrome' as const
+    }
+  }
+
+  match = RE_FIREFOX_STACKTRACE.exec(line)
+  if (match) {
+    let [input, varName, body, line, column] = match
+    // strip eval & function
+
+    if (/> eval|Function/.test(body)) {
+      body = body.replace(
+        / line (\d+)(?: > eval line \d+|(?: > eval|Function))*/,
+        ':$1'
+      )
+      const lcMatch = /(.*):(\d+)/.exec(body)
+      if (lcMatch) {
+        body = lcMatch[1]
+        line = lcMatch[2]
+        column = ''
+      } else {
+        // can't extract column
+        return {
+          input
+        }
+      }
+    }
+
+    return {
+      input: input,
+      varName: varName,
+      url: body,
+      line: Number(line),
+      column: Number(column),
+      vendor: 'firefox' as const
+    }
+  }
+
+  return {
+    input: line
+  }
+}
+
+const extractSourceMapData = (fileContent: string) => {
+  const re = /[#@]\ssourceMappingURL=\s*(\S+)/gm
+  let match: RegExpExecArray | null = null
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    const next = re.exec(fileContent)
+    if (next == null) {
+      break
+    }
+    match = next
+  }
+
+  if (!(match && match[0])) {
+    return null
+  }
+
+  return match[1]
+}
+
+const getSourceMapForFile = async (
+  fileName: string
+): Promise<{
+  sourceMapConsumer?: SourceMapConsumer
+  baseSource: string
+}> => {
+  const source = await (await fetch(fileName)).text()
```

Comment:
Pretty clever way to get the source map for the file

Though if we consider that we're going to fetch something from vite server anyway, what about sending the error to the server for normalization instead (e.g. through the WS link: `socket.send({ type: 'runtime-error', error})`) ?
That'd avoid duplicating the work (normalizing errors in vite server & doing it also in vite client) and would also limit the use of dependencies in the client.

My 2c

Reply:
Yea i agree, it would make more sense to use the websocket connection and move the processing over to the server.